### PR TITLE
Support optional client/serverError? in test hints

### PR DESCRIPTION
**Pull request by vdimir**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* `clickhouse client --testmode`: Add test hints for optional errors: `clientError?` / `serverError?`
* `clickhouse client --testmode`:  Allow not to specify exact error code in test hint

### CODE REVIEW

**Comment by vdimir** [comment]
Maybe convert current test into `sh` test, to check that errors that should be thrown are not suppressed?

File: tests/queries/0_stateless/02029_optional_error.sql [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+CREATE TABLE t ( a Int32 ) ENGINE = Memory;
+
+INSERT INTO t VALUES (abs(functionThatDoesNotExists(42))); -- { clientError? UNKNOWN_FUNCTION }
+INSERT INTO t VALUES (abs(functionThatDoesNotExists(42))); -- { clientError? }
+INSERT INTO t VALUES (abs(functionThatDoesNotExists(42))); -- { clientError }
+INSERT INTO t VALUES (1, 2); -- { clientError? SYNTAX_ERROR }
+INSERT INTO t VALUES (1, 2); -- { clientError? }
+INSERT INTO t VALUES (1, 2); -- { clientError }
+INSERT INTO t VALUES (1); -- { clientError? UNKNOWN_FUNCTION }
+INSERT INTO t VALUES (1); -- { clientError? SYNTAX_ERROR }
```
Comment (kssenii) [inline tests/queries/0_stateless/02029_optional_error.sql:12]: But it seemed that `?` means that we can either expect no error or exactly error after `?`. Why we allow two different errors?

**Comment by filimonov** [comment]
Why/when that non-determenism is needed? Tests usually should be very determenistic

**Comment by vdimir** [comment]
> Why/when that non-determenism is needed? Tests usually should be very determenistic

On this commit [`c661c4c` (#29075)](https://github.com/ClickHouse/ClickHouse/pull/29075/commits/c661c4cd6bf8368cdd61dd10c12e28739c181ba2) test failed, because no overflow occurred in CI, but in local run it happened, I guess it depends on machine.

Errors without specified code can be useful for development, for instance, when developing test and changing code we can not to care about exact code and specify it later when all finished.

File: tests/queries/0_stateless/02029_optional_error.sql (hunk 1 above)
Comment (vdimir) [reply tests/queries/0_stateless/02029_optional_error.sql:12]: Yes, this query are succeed, so both of codes will be ignored and do not affect anything.

**Review by kssenii — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Please rebase.

**Comment by vdimir** [comment]
Actually feature is controversial

**Pull request closed without merging**